lib/server/controllers/file_controller.rb

```diff
@@ -72,36 +74,115 @@ def rename
 
   def copy
     require_param(:new_file_path)
+
+    # Keep this here to verify the bucket is valid,
+    # since source bucket name is part of the path
     bucket = require_bucket
 
-    file = Metis::File.from_path(bucket, @params[:file_path])
+    raise Etna::Forbidden, "Cannot edit project #{@params[:project_name]}" unless @user.can_edit?(@params[:project_name])
```

Comment:
Hmm, should the route just be `can_edit? :project_name` instead of this? I think I initially thought: they only need to be able to read the file in the route path, not edit it - the edit permissions are for the write destination. If they are in the same project then the `can_edit?` permission should cover this. If not, the route should be `can_view?` and we should check if they can edit `new_project_name`. Since we don't support the latter at the moment it's probably easiest to just make the route `can_edit?`

Reply:
Oh yeah, I was wondering about this for the source file versus the destination file. Makes sense to just add it to the route definition, since we don't support changing projects at the moment.